client/src/main/java/io/pravega/client/security/auth/JwtTokenHandlingStrategy.java

```diff
@@ -0,0 +1,139 @@
+import io.pravega.common.concurrent.Futures;
+import io.pravega.common.util.ConfigurationOptionsExtractor;
+import lombok.AccessLevel;
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+
+import java.time.Instant;
+import java.util.Base64;
+import java.util.concurrent.atomic.AtomicReference;
+
+@Slf4j
+public class JwtTokenHandlingStrategy implements DelegationTokenHandlingStrategy {
+
+    /**
+     * Represents the threshold for triggering delegation token refresh.
+     */
+    @VisibleForTesting
+    static final int DEFAULT_REFRESH_THRESHOLD = 5;
+
+    @Getter(AccessLevel.PROTECTED)
+    private static final int REFRESH_THRESHOLD;
+
+    /**
+     * The Controller gRPC client used for interacting with the server.
+     */
+    private final Controller controllerClient;
+
+    private final String scopeName;
+
+    private final String streamName;
+
+    @Getter(AccessLevel.PROTECTED)
+    private AtomicReference<DelegationToken> delegationToken = new AtomicReference<>();
+
+    static {
+        REFRESH_THRESHOLD = ConfigurationOptionsExtractor.extractInt(
+                "pravega.client.auth.token-refresh.threshold",
+                "pravega_client_auth_token-refresh.threshold",
+                DEFAULT_REFRESH_THRESHOLD);
+    }
+
+    JwtTokenHandlingStrategy(Controller controllerClient, String scopeName, String streamName) {
+        Exceptions.checkNotNullOrEmpty(scopeName, "scopeName");
+        Preconditions.checkNotNull(controllerClient, "controllerClient is null");
+        Exceptions.checkNotNullOrEmpty(streamName, "streamName");
+
+        this.scopeName = scopeName;
+        this.streamName = streamName;
+        this.controllerClient = controllerClient;
+    }
+
+    public JwtTokenHandlingStrategy(String token, Controller controllerClient, String scopeName, String streamName) {
+        Exceptions.checkNotNullOrEmpty(token, "token");
+        Exceptions.checkNotNullOrEmpty(scopeName, "scopeName");
+        Preconditions.checkNotNull(controllerClient, "controllerClient is null");
+        Exceptions.checkNotNullOrEmpty(streamName, "streamName");
+
+        Long expTime = extractExpirationTime(token);
+        delegationToken.set(new DelegationToken(token, expTime));
+        this.scopeName = scopeName;
+        this.streamName = streamName;
+        this.controllerClient = controllerClient;
+    }
+
+    @Override
+    public String retrieveToken() {
+        if (isTokenNearingExpiry()) {
+            log.info("Token is nearing expiry, so refreshing it");
+            return refreshToken();
+        } else {
+            return delegationToken.get().getValue();
+        }
+    }
+
+    @Override
+    public String refreshToken() {
+        resetToken(newToken());
+        return delegationToken.get().getValue();
+    }
+
+    protected String newToken() {
+        return Futures.getAndHandleExceptions(
+                controllerClient.getOrRefreshDelegationTokenFor(scopeName, streamName), RuntimeException::new);
```

Comment:
SHould this have retries? Perhaps look at `Retry`?

Reply:
I think we should add it if needed later. I haven't encountered a scenario in my testing so far where retry was needed. I'll defer to your opinion though. Please let me know.

Reply:
You should have retries. Calls over the wire are bound to fail for whatever reasons. Having an exp-backoff retry will save you a lot of trouble in the future.

Reply:
Sorry, I take it back. There are retries built into that method.